_posts/2019-09-20-lambda-to-openfaas.md

```diff
@@ -0,0 +1,282 @@
+
+Let's have a look at the `shortener` function which creates and stores the shortened URLs:
+
+```javascript
+const AWS = require('aws-sdk');
+
+const shortDomain = 'https://checkthisnow.net/'
+let docClient = new AWS.DynamoDB.DocumentClient();
+
+exports.handler = function (event, context, callback) {
+    let originalUrl = JSON.parse(event.body).url || '';
+
+    checkExists(originalUrl, function (exist) {
+        if (!exist) {
+            createAndStorePath(originalUrl, callback);
+        } else {
+            let shortenedURL = shortDomain + exist.token;
+            let response = {
+                isBase64Encoded: false,
+                statusCode: 200,
+                headers: {},
+                body: JSON.stringify({
+                    url: shortenedURL
+                })
+            };
+            callback(null, response);
+        }
+    });
+}
+
+function checkExists(url, callback) {
+    docClient.query(params, function (err, data) {
+        ...
+    });
+}
+
+function createAndStorePath(originalUrl, callback) {
+    let path = generateToken();
+    let shortenedURL = shortDomain + path;
+    storeURL(path, originalUrl, function (err) {
+        ...
+    });
+}
+
+function generateToken(path = '') {
+    ...
+}
+
+function storeURL(token, originalUrl, callback) {
+    docClient.put(params, function (err, data) {
+        ...
+    });
+}
+```
+
+A couple of things to note for this function code:
+- The input comes as the `event` object and must be parsed before accessing the properties
+- The DynamoDB `DocumentClient` doesn't require any additional parameters to connect
+  - This is because AWS maps the execution environment to the account and IAM role when you create the function
+- The response is an AWS defined object where the body must be converted to a `string`
+- Not obvious from the code, but the `aws-sdk` does not need to be included in the `package.json` `dependencies` as it is included when the code is re-packaged for execution in Lambda.
+
+There is also the additional configuration that is required in the AWS API Gateway:
+- Creating each of the routes and methods
+- Ensuring the request properties are configured properly to allow access to the request body
+- In the case of the `redirecter` function: 
+  - Creating the "Resource" variable `/{token}`
+  - Adding each of the HTTP response codes
+  - Mapping the response to the appropriate header values
+
+All of this setup and configuration outside of the function code means that the function, while seemingly simple, can only be run within the AWS Lambda environment, and only after several other steps are performed.
+
+## OpenFaaS
+
+#### Create the New Function
+Pull the Node.js + Express.js template to build from it:
+```shell
+$ faas-cli template store pull node10-express
+```
+Now, create the new "shortener" function:
+```shell
+$ faas-cli new shortener --lang node10-express
```

Comment:
should we use `--prefix` here, or set the env var?  It'll save mentioning the image prefix in the note later.

Reply:
+1

Reply:
Good idea! I added a bit about using the `DOCKER_USER` env var that makes the rest of the commands shorter